File: src/main/java/fr/tikione/c2e/service/web/scrap/AbstractScraper.java
package fr.tikione.c2e.service.web.scrap;

import fr.tikione.c2e.model.web.Article;
import fr.tikione.c2e.service.web.AbstractReader;
import lombok.SneakyThrows;
import org.jsoup.nodes.Document;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class AbstractScraper extends AbstractReader implements CPCScraperService {
    
    public int rate(List<Article> articles) {
        return articles == null ? -1 : articles.stream().mapToInt(Article::rate).sum();
    }
    
    @Override
    @SneakyThrows
    public Map<Integer, List<Article>> extractArticles(Document doc) {
        Map<Integer, List<Article>> score = new HashMap<>();
        
        try {
            List<Article> news = extractNews(doc);
            score.put(rate(news), news);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            List<Article> singleNews = extractSingleNews(doc);
            score.put(rate(singleNews), singleNews);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            List<Article> tests = extractTests(doc);
            score.put(rate(tests), tests);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            List<Article> shortTests = extractShortTests(doc);
            score.put(rate(shortTests), shortTests);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            List<Article> plumePudding = extractPlumePudding(doc);
            score.put(rate(plumePudding), plumePudding);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            List<Article> coming = extractComing(doc);
            score.put(rate(coming), coming);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            List<Article> underConstruction = extractUnderConstruction(doc);
            score.put(rate(underConstruction), underConstruction);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            List<Article> techno = extractTechno(doc);
            score.put(rate(techno), techno);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            List<Article> study = extractStudy(doc);
            score.put(rate(study), study);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        try {
            List<Article> everythingElse = extractEverythingElse(doc);
            score.put(rate(everythingElse), everythingElse);
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        return score;
    }
    
    @Override
    public List<Article> extractSingleNews(Document doc) {
        return Collections.emptyList();
    }
    
    @Override
    public List<Article> extractNews(Document doc) {
        return Collections.emptyList();
    }
    
    @Override
    public List<Article> extractTests(Document doc) {
        return Collections.emptyList();
    }
    
    @Override
    public List<Article> extractShortTests(Document doc) {
        return Collections.emptyList();
    }
    
    @Override
    public List<Article> extractPlumePudding(Document doc) {
        return Collections.emptyList();
    }
    
    @Override
    public List<Article> extractComing(Document doc) {
        return Collections.emptyList();
    }
    
    @Override
    public List<Article> extractUnderConstruction(Document doc) {
        return Collections.emptyList();
    }
    
    @Override
    public List<Article> extractTechno(Document doc) {
        return Collections.emptyList();
    }
    
    @Override
    public List<Article> extractStudy(Document doc) {
        return Collections.emptyList();
    }
    
    @Override
    public List<Article> extractEverythingElse(Document doc) {
        return Collections.emptyList();
    }
}

File: src/main/kotlin/fr/tikione/c2e/service/web/scrap/AbstractScraper.kt
package fr.tikione.c2e.service.web.scrap

import fr.tikione.c2e.model.web.Article
import fr.tikione.c2e.service.web.AbstractReader
import org.jsoup.nodes.Document

abstract class AbstractScraper : AbstractReader(), CPCScraperService {

    private fun rate(articles: List<Article>?): Int {
        if (articles == null) {
            return -1
        }
        return articles.sumBy { it.rate() }
    }



    override fun extractBestArticles(doc: Document): List<Article> {
        val extractions = listOf(
                safe { extractNews(doc) },
                safe { extractSingleNews(doc) },
                safe { extractPlumePudding(doc) },
                safe { extractShortTests(doc) },
                safe { extractTests(doc) },
                safe { extractComing(doc) },
                safe { extractUnderConstruction(doc) },
                safe { extractTechno(doc) },
                safe { extractStudy(doc) },
                safe { extractEverythingElse(doc) }
        )

        return extractions.map { Pair(rate(it), it ?: emptyList()) }
                //sort by highest score (first) and return the articles (second)
                .sortedByDescending { it.first }.map { it.second }[0]
    }

    private inline fun <T> safe(function: () -> T): T? {
        return try {
            function()
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    override fun extractSingleNews(doc: Document): List<Article> {
        return emptyList()
    }

    override fun extractNews(doc: Document): List<Article> {
        return emptyList()
    }

    override fun extractTests(doc: Document): List<Article> {
        return emptyList()
    }

    override fun extractShortTests(doc: Document): List<Article> {
        return emptyList()
    }

    override fun extractPlumePudding(doc: Document): List<Article> {
        return emptyList()
    }

    override fun extractComing(doc: Document): List<Article> {
        return emptyList()
    }

    override fun extractUnderConstruction(doc: Document): List<Article> {
        return emptyList()
    }

    override fun extractTechno(doc: Document): List<Article> {
        return emptyList()
    }

    override fun extractStudy(doc: Document): List<Article> {
        return emptyList()
    }

    override fun extractEverythingElse(doc: Document): List<Article> {
        return emptyList()
    }
}
